In [383]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score

In [384]:
df = pd.read_csv('/Users/anastasia/Downloads/train_with_new_features.csv')
df = df.drop(columns=['id', 'address', 'address_rus'])
df = df.dropna()
df

,atm_group,lat,long,target,schools_nearby,supermarket_nearby,mall_nearby,bar_nearby,cafe_nearby,restaurant_nearby,police_nearby,post_office_nearby,place_of_worship_nearby,university_nearby,cinema_nearby,casino_nearby,nightclub_nearby,atm_nearby,population
0,32,46.940995,142.738319,0.011500,2,3,2,0,0,0,0,0,0,0,0,0,0,0.0,180467
1,32,46.937353,142.753348,0.029710,3,2,5,1,4,2,0,1,1,0,0,0,0,1.0,180467
2,32,46.959413,142.741113,0.009540,4,2,2,2,7,2,0,0,0,1,0,0,1,3.0,180467
3,32,55.805837,37.515164,-0.094035,4,5,2,4,29,12,3,5,1,0,0,0,0,13.0,13104177
4,32,69.343541,88.211228,0.079277,7,2,0,0,6,1,0,2,2,1,1,0,0,2.0,179554
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6256,8083,62.045221,129.748851,0.207916,4,2,3,0,3,0,0,1,0,1,0,0,0,1.0,361154
6257,8083,53.769459,87.117251,0.062220,1,0,0,0,1,0,2,1,0,0,0,0,0,0.0,533565
6258,8083,55.766212,37.380875,0.108743,1,1,1,0,1,0,0,0,0,0,1,0,0,0.0,201
6259,8083,56.838223,86.640829,0.148938,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,5137


In [385]:
X_train = df.drop(columns=['target'])
y_train = df['target']

In [386]:
df_test = pd.read_csv('/Users/anastasia/Downloads/test_with_new_features.csv')
df_test = df_test.drop(columns=['Unnamed: 0', 'id', 'address', 'address_rus'])
df_test = df_test.dropna()
df_test

,atm_group,lat,long,police_nearby,post_office_nearby,place_of_worship_nearby,schools_nearby,bar_nearby,cafe_nearby,restaurant_nearby,cinema_nearby,casino_nearby,nightclub_nearby,university_nearby,supermarket_nearby,mall_nearby,atm_nearby,population,target
0,5478.0,53.757824,87.159786,2.0,2.0,0,8,0,1,4,0,0,0,0,2,0,3,533565.0,-0.037550
1,3185.5,55.763346,37.551241,0.0,4.0,3,2,0,14,10,0,0,0,0,1,0,0,13274285.0,-0.039498
2,3185.5,52.918598,158.514038,0.0,1.0,1,3,1,2,0,0,0,0,0,0,0,4,22386.0,-0.015270
4,1942.0,55.179216,61.297729,0.0,3.0,0,3,0,1,0,0,0,0,0,5,4,3,101285.0,0.000239
5,5478.0,54.971827,73.411646,1.0,1.0,0,1,0,12,0,0,0,0,2,2,1,3,1101367.0,-0.029048
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2499,5478.0,55.886488,37.656829,1.0,2.0,1,5,1,4,2,1,0,0,0,5,2,14,466.0,-0.026097
2500,5478.0,43.438310,39.912211,0.0,1.0,1,0,0,8,2,0,0,0,0,3,1,2,115000.0,-0.052060
2501,8083.0,55.748275,37.583571,2.0,5.0,3,2,6,21,48,0,0,1,2,4,2,7,13274285.0,0.183749
2502,1942.0,55.910992,37.748466,0.0,0.0,0,1,0,5,1,0,0,0,0,2,1,0,225299.0,0.037698


In [387]:
X_test = df_test.drop(columns=['target'])
Y_test = df_test['target']

In [388]:
column_names = X_train.columns
column_names

Index(['atm_group', 'lat', 'long', 'schools_nearby', 'supermarket_nearby',
       'mall_nearby', 'bar_nearby', 'cafe_nearby', 'restaurant_nearby',
       'police_nearby', 'post_office_nearby', 'place_of_worship_nearby',
       'university_nearby', 'cinema_nearby', 'casino_nearby',
       'nightclub_nearby', 'atm_nearby', 'population'],
      dtype='object')

In [389]:
X_train['population'] = X_train['population'].str.replace("\xa0", "").astype(float)

## LinearRegression

In [ ]:
model = LinearRegression()
model.fit(X_train.astype(float), y_train)
y_pred = model.predict(X_test[column_names])

/Users/anastasia/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: divide by zero encountered in matmul
  return X @ coef_ + self.intercept_
/Users/anastasia/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: overflow encountered in matmul
  return X @ coef_ + self.intercept_
/Users/anastasia/Library/Python/3.9/lib/python/site-packages/sklearn/linear_model/_base.py:279: RuntimeWarning: invalid value encountered in matmul
  return X @ coef_ + self.intercept_


In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error, mean_absolute_error
from sklearn.metrics import r2_score
import numpy as np

mse_model = mean_squared_error(Y_test, y_pred)
mape_model = mean_absolute_percentage_error(Y_test, y_pred)
mae_model = mean_absolute_error(Y_test, y_pred)
r2_model = r2_score(Y_test, y_pred)
print(f"MSE: {mse_model:.4f}")
print(f"MAPE: {mape_model:.4f}")
print(f"MAE: {mae_model:.4f}")
print(f"R² (коэффициент детерминации): {r2_model:.6f}")
print(f"R² в процентах: {r2_model * 100:.2f}%")

MSE: 0.0030
MAPE: 12.3437
MAE: 0.0485
R² (коэффициент детерминации): 0.374317
R² в процентах: 37.43%


##### Средняя квадратичная ошибка составляет 0.0030 квадратных единиц. Средняя абсолютная процентная ошибка составляет 12.34%. Средняя абсолютная ошибка составляет 0.0485 единиц. Модель объясняет только 37.43% дисперсии целевой переменной

## SVM

In [ ]:
from sklearn import svm
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test[column_names])
model = svm.SVR(kernel='rbf', C=1.0, gamma='scale')
model.fit(X_train_scaled, y_train)

SVR()

In [ ]:
y_pred = model.predict(X_test[column_names])

/Users/anastasia/Library/Python/3.9/lib/python/site-packages/sklearn/utils/validation.py:2732: UserWarning: X has feature names, but SVR was fitted without feature names
  warnings.warn(


In [ ]:
mse_svm = mean_squared_error(Y_test, y_pred)
mape_svm = mean_absolute_percentage_error(Y_test, y_pred)
mae_svm = mean_absolute_error(Y_test, y_pred)
r2_svm = r2_score(Y_test, y_pred)
print(f"MSE: {mse_svm:.4f}")
print(f"MAPE: {mape_svm:.4f}")
print(f"MAE: {mae_svm:.4f}")
print(f"R² (коэффициент детерминации): {r2_svm:.6f}")
print(f"R² в процентах: {r2_svm * 100:.2f}%")

MSE: 0.0073
MAPE: 21.0175
MAE: 0.0812
R² (коэффициент детерминации): -0.500241
R² в процентах: -50.02%


##### Средняя квадратичная ошибка составляет 0.0073 квадратных единиц. Средняя абсолютная процентная ошибка составляет 21.0175%. Средняя абсолютная ошибка составляет 0.0812 единиц. Модель объясняет только -50.02% дисперсии целевой переменной.

##### То есть модель дает результат сильно хуже, чем простое предсказание средним значением

## DecisionTreeRegressor

In [ ]:
from sklearn.tree import DecisionTreeRegressor
dt_reg = DecisionTreeRegressor(
    max_depth=5,            # Ограничение глубины
    min_samples_split=10,   # Минимум samples для разделения
    min_samples_leaf=5,     # Минимум samples в листе
    random_state=42
)
dt_reg.fit(X_train, y_train)
y_pred = dt_reg.predict(X_test[column_names])

In [ ]:
mse_dt_reg = mean_squared_error(Y_test, y_pred)
mape_dt_reg = mean_absolute_percentage_error(Y_test, y_pred)
mae_dt_reg = mean_absolute_error(Y_test, y_pred)
r2_dt_reg = r2_score(Y_test, y_pred)
print(f"MSE: {mse_dt_reg:.4f}")
print(f"MAPE: {mape_dt_reg:.4f}")
print(f"MAE: {mae_dt_reg:.4f}")
print(f"R² (коэффициент детерминации): {r2_dt_reg:.6f}")
print(f"R² в процентах: {r2_dt_reg * 100:.2f}%")

MSE: 0.0003
MAPE: 13.7205
MAE: 0.0144
R² (коэффициент детерминации): 0.927944
R² в процентах: 92.79%


##### Использование рещающего дерева сильно улучшило результат: MSE, MAPE и MAE значительно упали, а доля объясненной дисперсии увеличилась до 92.79

## RandomForestRegressor

In [ ]:
from sklearn.ensemble import RandomForestRegressor
regr = RandomForestRegressor(max_depth=5, random_state=42)
regr.fit(X_train, y_train)
y_pred = regr.predict(X_test[column_names])

In [ ]:
mse_regr = mean_squared_error(Y_test, y_pred)
mape_regr = mean_absolute_percentage_error(Y_test, y_pred)
mae_regr = mean_absolute_error(Y_test, y_pred)
r2_regr = r2_score(Y_test, y_pred)
print(f"MSE: {mse_regr:.4f}")
print(f"MAPE: {mape_regr:.4f}")
print(f"MAE: {mae_regr:.4f}")
print(f"R² (коэффициент детерминации): {r2_regr:.6f}")
print(f"R² в процентах: {r2_regr * 100:.2f}%")

MSE: 0.0002
MAPE: 12.1383
MAE: 0.0125
R² (коэффициент детерминации): 0.948801
R² в процентах: 94.88%


## GradientBoostingRegressor

In [ ]:
from sklearn.ensemble import GradientBoostingRegressor
gbr = GradientBoostingRegressor(
    n_estimators=100,       # Количество деревьев
    learning_rate=0.1,      # Скорость обучения
    max_depth=3,            # Глубина каждого дерева
    min_samples_split=2,
    min_samples_leaf=1,
    random_state=42
)
gbr.fit(X_train, y_train)
y_pred = gbr.predict(X_test[column_names])

In [ ]:
mse_gbr = mean_squared_error(Y_test, y_pred)
mape_gbr = mean_absolute_percentage_error(Y_test, y_pred)
mae_gbr = mean_absolute_error(Y_test, y_pred)
r2_gbr = r2_score(Y_test, y_pred)
print(f"MSE: {mse_gbr:.4f}")
print(f"MAPE: {mape_gbr:.4f}")
print(f"MAE: {mae_gbr:.4f}")
print(f"R² (коэффициент детерминации): {r2_gbr:.6f}")
print(f"R² в процентах: {r2_gbr * 100:.2f}%")

MSE: 0.0002
MAPE: 4.3321
MAE: 0.0111
R² (коэффициент детерминации): 0.957697
R² в процентах: 95.77%


## GridSearchCV

Попробуем подобрать лучшие гиперпараметры для KNN

In [359]:
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsRegressor

In [360]:
parameters = {
    'n_neighbors': [3, 5, 7, 9, 11],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan']
}

base_reg = KNeighborsRegressor()
reg_cv = GridSearchCV(base_reg, parameters)
reg_cv.fit(X_train, y_train)

GridSearchCV(estimator=KNeighborsRegressor(),
             param_grid={'metric': ['euclidean', 'manhattan'],
                         'n_neighbors': [3, 5, 7, 9, 11],
                         'weights': ['uniform', 'distance']})

In [361]:
reg_cv.best_estimator_

KNeighborsRegressor(metric='manhattan', n_neighbors=11, weights='distance')

In [362]:
y_pred = reg_cv.predict(X_test[column_names])

In [363]:
mse_knn = mean_squared_error(Y_test, y_pred)
mape_knn = mean_absolute_percentage_error(Y_test, y_pred)
mae_knn = mean_absolute_error(Y_test, y_pred)
r2_knn = r2_score(Y_test, y_pred)
print(f"MSE: {mse_knn:.4f}")
print(f"MAPE: {mape_knn:.4f}")
print(f"MAE: {mae_knn:.4f}")
print(f"R² (коэффициент детерминации): {r2_knn:.6f}")
print(f"R² в процентах: {r2_knn * 100:.2f}%")

MSE: 0.0011
MAPE: 12.9095
MAE: 0.0249
R² (коэффициент детерминации): 0.782409
R² в процентах: 78.24%


In [ ]:
pd.DataFrame({
    "model": ["LinearRegression", "svm", "DecisionTree", "RandomForest", "GradientBoosting", "KNN"],
    "mse": [mse_model, mse_svm, mse_dt_reg, mse_regr, mse_gbr, mse_knn],
    "mape": [mape_model, mape_svm, mape_dt_reg, mape_regr, mape_gbr, mape_knn],
    "mae": [mae_model, mae_svm, mae_dt_reg, mae_regr, mae_gbr, mae_knn],
    "r2": [r2_model, r2_svm, r2_dt_reg, r2_regr, r2_gbr, r2_knn]
}).style.set_caption('Сравнение моделей машинного обучения')\
.set_table_attributes('class="dataframe" style="border-collapse: collapse; width: 100%;"')

,model,mse,mape,mae,r2
0,LinearRegression,0.003036,12.343706,0.048473,0.374317
1,svm,0.007279,21.017480,0.081154,-0.500241
2,DecisionTree,0.000350,13.720472,0.014352,0.927944
3,RandomForest,0.000248,12.138253,0.012530,0.948801
4,GradientBoosting,0.000205,4.332113,0.011130,0.957697
5,KNN,0.001056,12.909457,0.024877,0.782409


### Итого мы перебрали различные модели и попробовали подобрать наилучшие гиперпараметры. Наилучший результат демонстрирует градиентный бустринг, имея наивысший R2 и наименьшую ошибку. Эти выводы нам помогут при дальнейшей работе и улучшения существующих моделей 